### ISSUE TITLE
sharding_constraint and shard_map require relowering for meshes with same shape and axis but different devices.

### ISSUE THREAD

**Issue by Vikisix** [google/jax#19638]
### Description

We have the following code:
```python
class Foo:
  def __init__(self, mesh: jax.sharding.Mesh):
    self.foo = lambda *a, __f=self.foo, **k: __f(*a, **k) # trick to keep the function id stable
    self.foo_2 = lambda *a, __f=self.foo_2, **k: __f(*a, **k) # trick to keep the function id stable
    self.mesh = mesh

  def foo(self, x):
    with self.mesh:
      @jax.jit
      def _foo(x):
        return x + 1
      print('trace')
      return shard_map.shard_map(_foo, self.mesh, in_specs=jax.sharding.PartitionSpec('x'), out_specs=jax.sharding.PartitionSpec('x'), check_rep=False)(x)

  def foo_2(self, x):
    with self.mesh:
      print('trace')
      return x + jnp.ones(tuple(self.mesh.shape.values()), dtype=jnp.float32)
```
In one case we are passing the mesh to `shard_map` in one case we are using the shape for doing some kind of operations.
We have two meshes with the **same shape** and **same axis_name**.
```python
mesh_1 = jax.sharding.Mesh(np.asarray(jax.devices()[:4]).reshape(2, 2), axis_names=['x', 'y'])
mesh_2 = jax.sharding.Mesh(np.asarray(jax.devices()[4:]).reshape(2, 2), axis_names=['x', 'y'])
named_sharding_1 = jax.sharding.NamedSharding(mesh_1, jax.sharding.PartitionSpec('x'))
named_sharding_2 = jax.sharding.NamedSharding(mesh_2, jax.sharding.PartitionSpec('x'))
```
When we execute the following code:
```python
foo = Foo(mesh_1)
arr_1 = ... # array on mesh_1
j_f_1 = jax.jit(foo.foo_2 ,in_shardings=named_sharding_1, out_shardings=named_sharding_1)
l_f_1 = j_f_1.lower(arr_1) # It prints trace
c_f_1 = l_f_1.compile()
arr_1 = c_f_1(arr_1)
arr_1 = ... # array on mesh_2
j_f_2 = jax.jit(foo.foo_2 ,in_shardings=named_sharding_2, out_shardings=named_sharding_2)
l_f_2 = j_f_2.lower(arr_2) # No traces happens here.
c_f_2 = l_f_1.compile()
arr_2 = c_f_1(arr_2) # works fine without problem
```
However if we execute the same code for `foo.foo`:
```
foo = Foo(mesh_1)
arr_1 = ... # array on mesh_1
j_f_1 = jax.jit(foo.foo ,in_shardings=named_sharding_1, out_shardings=named_sharding_1)
l_f_1 = j_f_1.lower(arr_1) # It prints trace
c_f_1 = l_f_1.compile() 
arr_1 = c_f_1(arr_1) # Works fine
arr_1 = ... # array on mesh_2
j_f_2 = jax.jit(foo.foo_2 ,in_shardings=named_sharding_2, out_shardings=named_sharding_2)
l_f_2 = j_f_2.lower(arr_2) # Exception!
```
We end up with `DeviceAssignmentMismatch` since the two meshes has different devices id.

### What jax/jaxlib version are you using?

HEAD

### Which accelerator(s) are you using?

TPU

**Comment by Vikisix** [comment]
This is a full reproducible code:
```python
import jax
from jax import numpy as jnp
import numpy as np
from jax.experimental import shard_map
import math
half_device = jax.device_count() // 2
mesh_1 = jax.sharding.Mesh(np.asarray(jax.devices()[:half_device]).reshape(2, 2), axis_names=['x', 'y'])
mesh_2 = jax.sharding.Mesh(np.asarray(jax.devices()[half_device:]).reshape(2, 2), axis_names=['x', 'y'])
named_sharding_1 = jax.sharding.NamedSharding(mesh_1, jax.sharding.PartitionSpec('x'))
named_sharding_2 = jax.sharding.NamedSharding(mesh_2, jax.sharding.PartitionSpec('x'))



input_shape = (2, 2)
global_input_data = np.arange(math.prod(input_shape), dtype=np.float32).reshape(input_shape)

def cb(index):
 return global_input_data[index]

arr_1 = jax.make_array_from_callback(input_shape, named_sharding_1, cb)
arr_2 = jax.make_array_from_callback(input_shape, named_sharding_2, cb)

def foo(x):
    @jax.jit
    def _foo(x):
      return x + 1
    print('trace')
    return shard_map.shard_map(_foo, mesh_1, in_specs=jax.sharding.PartitionSpec('x'), out_specs=jax.sharding.PartitionSpec('x'), check_rep=False)(x)

def foo_2(x):
    print('trace')
    return x + jnp.ones(tuple(mesh_1.shape.values()), dtype=jnp.float32)

print('foo_2')
print("******")
arr_1 = jax.make_array_from_callback(input_shape, named_sharding_1, cb)
shapes_1 = jax.eval_shape(foo_2, arr_1)
j_f_1 = jax.jit(foo_2, in_shardings=named_sharding_1, out_shardings=named_sharding_1, donate_argnums=[0])
l_f_1 = j_f_1.lower(arr_1) # tracing
c_f_1 = l_f_1.compile()
arr_1 = c_f_1(arr_1) # works fine
print(arr_1)
print("******")
arr_2 = jax.make_array_from_callback(input_shape, named_sharding_2, cb)
j_f_2 = jax.jit(foo_2, in_shardings=named_sharding_2, out_shardings=named_sharding_2, donate_argnums=[0])
l_f_2 = j_f_2.lower(shapes_1) # No trace
c_f_2 = l_f_2.compile() 
arr_2 = c_f_2(arr_2)
print(arr_2)  # works fine 
print('foo')
print("******")
shapes_1 = jax.eval_shape(foo, arr_1)
arr_1 = jax.make_array_from_callback(input_shape, named_sharding_1, cb)
j_f_1 = jax.jit(foo, in_shardings=named_sharding_1, out_shardings=named_sharding_1, donate_argnums=[0])
l_f_1 = j_f_1.lower(arr_1) # tracing
c_f_1 = l_f_1.compile()
arr_1 = c_f_1(arr_1) # works fine
print(arr_1)
print("******")
arr_2 = jax.make_array_from_callback(input_shape, named_sharding_2, cb)
j_f_2 = jax.jit(foo, in_shardings=named_sharding_2, out_shardings=named_sharding_2, donate_argnums=[0])
l_f_2 = j_f_2.lower(shapes_1) # Exception: Received incompatible devices for jitted computation if arr_2 is used or DeviceAssignmentMismatch if shapes are used
```
